### PR TITLE
update for svd2rust v0.26.0

### PR DESCRIPTION
**Pull request by almindor**
Update crate for use with `svd2rust` version `0.26.0`

* move src/common into src directly to avoid need for pre-generated lib.rs
* remove `update.sh` hacks
* add `critical-section` support via the `riscv` crate's `critical-section-single-hart` feature

### CODE REVIEW

**Comment by almindor** [comment]
@Disasm it seems `cargo` got updated in `1.60` to support same name for a feature and dependency. svd2rust now generates the `critical-section` config/feature, I guess it means we have to bump to 1.60?

File: update.sh [hunk 1]
```diff
@@ -3,22 +3,10 @@ set -x
 set -e
 
 # used svd2rust v0.19.0
```
Comment (9names) [inline update.sh:5]: This comment is out of date

File: CHANGELOG.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
 
 - Update `riscv` dependency to version 0.8
-- Update Minimum Supported Rust Version to 1.59
+- Update Minimum Supported Rust Version to 1.60
 - Regenerate code with `svd2rust` v0.19.0
```
Comment (9names) [inline CHANGELOG.md:21]: These lines are for the previous release, you should leave them as-is and put the new MSRV, riscv, and svd2rust versions under the unreleased heading.

**Comment by Disasm** [comment]
I don't think we need a separate `critical-section` feature. It can be removed and `critical-section` dependency can be made optional. Here is a possible Cargo.toml change which compiles with 1.59:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index 3c5c973..7f6a633 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,10 +10,11 @@ license = "ISC"
 rust-version = "1.59"
 
 [dependencies]
+critical-section = { version = "1.1.1", optional = true }
 vcell = "0.1.3"
-riscv = "0.10.1"
 
 [features]
+default = ["critical-section"]
 rt = []
 g002 = []
 
 ```

File: src/interrupt.rs [hunk 3]
```diff
@@ -107,9 +107,11 @@ pub enum Interrupt {
     #[doc = "52 - I2C0"]
     I2C0 = 52,
 }
+#[doc = r" TryFromInterruptError"]
```
Comment (romancardenas) [inline src/interrupt.rs:110]: This is fine, but the `interrupt!` macro at the bottom of the file should be reformatted. Currently is just impossible to read

File: src/lib.rs [hunk 4]
```diff
@@ -17,26 +18,668 @@ svd2rust release can be generated by cloning the svd2rust [repository], checking
...
 #![allow(non_snake_case)]
 #![no_std]
+use core::marker::PhantomData;
+use core::ops::Deref;
 #[allow(unused_imports)]
 use generic::*;
-#[doc = r"Common register and bit access and modify traits"]
+#[doc = "Common register and bit access and modify traits"]
 pub mod generic;
+#[cfg(feature = "rt")]
```
Comment (romancardenas) [inline src/lib.rs:27]: This is part of our latest contribution in `e310x-hal`. Once we merge this PR to `master`, we will probably need to clean the HAL.

File: src/lib.rs [hunk 5]
```diff
@@ -17,26 +18,668 @@ svd2rust release can be generated by cloning the svd2rust [repository], checking
...
+}
+#[doc(hidden)]
+pub union Vector {
+    pub _handler: unsafe extern "C" fn(),
+    pub _reserved: usize,
+}
+#[cfg(feature = "rt")]
+#[doc(hidden)]
+#[no_mangle]
+pub static __EXTERNAL_INTERRUPTS: [Vector; 53] = [
```
Comment (romancardenas) [inline src/lib.rs:90]: Same here, we must remove the `virq` thing from `e310x-hal`.

File: src/plic.rs [hunk 6]
```diff
@@ -2,36 +2,36 @@
 #[repr(C)]
 pub struct RegisterBlock {
```
Comment (romancardenas) [inline src/plic.rs:3]: As future work, we can consider removing this implementation and using the generic PLIC implementation I proposed in [`riscv`](https://github.com/rust-embedded/riscv/pull/125)

**Review by romancardenas — COMMENTED** [review]
Looks good to me. I noticed that we need to rework the `virq` feature for the HAL, as now this crate does half of the work.

**Comment by almindor** [comment]
> ```diff
> +critical-section = { version = "1.1.1", optional = true }
> ```

But that makes this locked out on the `Peripheral` no?

```rust
    #[cfg(feature = "critical-section")]
    #[inline]
    pub fn take() -> Option<Self> {
```

**Comment by 9names** [comment]
> But that makes this locked out on the Peripheral no?

You can't `take()` but you can still `steal()`
If you're providing safety some other way and don't want to use critical section this configuration allows that option.

**Comment by almindor** [comment]
> > But that makes this locked out on the Peripheral no?
> 
> You can't `take()` but you can still `steal()` If you're providing safety some other way and don't want to use critical section this configuration allows that option.

Maybe I misunderstand something but if we don't declare a `critical-section` feature in `Cargo.toml` won't it mean the users of this crate will have no way to enable it so they can have a `Peripherals::take` ?

**Comment by 9names** [comment]
Having an optional dependency defines a feature for you. 
https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

**Comment by almindor** [comment]
> Having an optional dependency defines a feature for you. https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

:facepalm: years of rust and here I am missing this :D

**Comment by almindor** [comment]
reverted to MSRV 1.59 with optional critical section feature. This should be "good to go" now.

**Review by romancardenas — APPROVED** [review]
LGTM. In fact, this version together with the PLIC peripheral will allow us to have RTIC apps running on E310x chips I hope!

**Pull request merged**